### ISSUE TITLE
 No implementation of method: :make-reader for leiningen-version

### ISSUE THREAD

**Issue by arichiardi** [boot-clj/boot-new#19]
Hello Sean,
I know this might or might not be a `boot-new` concern but I am going to report it here for discussion.

Sometimes templates use Leiningen-specific functions, like:
`[leiningen.core.main :as main :refer [leiningen-version]]`

<s>This function is very useful and big templates like `luminus` take advantage or that.
Maybe a solution could be to include Leiningen classes on the classpath in `boot-new`?
Otherwise `boot-new` cannot generate those "popular" templates.</s>

When I call it, I get:

```
        clojure.lang.ExceptionInfo: No implementation of method: :make-reader of protocol: #'clojure.java.io/IOFactory found for class: nil
    data: {:file "/tmp/boot.user2705218119118993821.clj", :line 15}
java.lang.IllegalArgumentException: No implementation of method: :make-reader of protocol: #'clojure.java.io/IOFactory found for class: nil
      clojure.core/-cache-protocol-fn  core_deftype.clj:  554
                 clojure.java.io/fn/G            io.clj:   69
               clojure.java.io/reader            io.clj:  102
                                  ...                        
leiningen.core.main/leiningen-version          main.clj:  321
leiningen.new.lambone/version-before?       lambone.clj:   34
        leiningen.new.lambone/lambone       lambone.clj:   59
                                  ...                        
                   clojure.core/apply          core.clj:  632
             boot.new-helpers/create*   new_helpers.clj:  105
              boot.new-helpers/create   new_helpers.clj:  118
                                  ...                        
             boot.new/eval74/fn/fn/fn           new.clj:   43
                  boot.core/run-tasks          core.clj:  943
                    boot.core/boot/fn          core.clj:  953
  clojure.core/binding-conveyor-fn/fn          core.clj: 1916
                                  ...                        
```

I was wondering if there is a way to identify that I am creating the template with `boot-new` and avoid the function call.

**Comment by seancorfield** [comment]
The Leiningen core library _is_ on the classpath and several templates rely on that and work just fine.

The problem with `leiningen-version` and some others functions is that they reach out to environment variables (set by the `lein` shell script) or the `project.clj` file itself or the non-core Leiningen files -- which `boot-new` cannot provide.

Specifically `leiningen-version` checks an environment variable (which Java code cannot set) and reads the `pom.properties` file out of Leiningen's `META-INF` folder tree (which is not part of the Leiningen core).

**Comment by seancorfield** [comment]
See https://github.com/technomancy/leiningen/blob/master/leiningen-core/src/leiningen/core/main.clj#L330-L337

**Comment by arichiardi** [comment]
Oh that explains why I have this error:

```
        clojure.lang.ExceptionInfo: No implementation of method: :make-reader of protocol: #'clojure.java.io/IOFactory found for class: nil
    data: {:file "/tmp/boot.user2705218119118993821.clj", :line 15}
```

The reader in is of course `nil` at that point...I wonder if I can detect if I am running `boot-new` and skip the call somehow.

**Comment by arichiardi** [comment]
I have updated the issue, can still be useful to know a workaround I guess.

**Comment by seancorfield** [comment]
As a hack, I could include a fake `pom.properties` for Leiningen in `boot-new`'s JAR that should allow `leiningen.core.main/leiningen-version` to be satisfied.

**Comment by arichiardi** [comment]
Yes I was thinking about that, but are you going to return a real lein version or like `boot-new` as string? This poses the issue of which `lein` version `boot-new` support I guess.

**Comment by seancorfield** [comment]
Yes, `boot-new` would "lock" the version of Leiningen it supported so if templates started requiring new features in Leiningen, I'd have to review them anyway and see if `boot-new` needed to be updated to support them (and bump the "supported Leiningen version" either way).

**Comment by arichiardi** [comment]
Yeah definitely true, it would be great if you could be 100% `lein` compatible.

**Comment by seancorfield** [comment]
Fixed in release 0.4.3 which just dropped on Clojars. I tested that I can generate a Luminus template.

**Comment by arichiardi** [comment]
Great! Thanks super quick ;)
